hazelcast/src/main/java/com/hazelcast/nio/tcp/nonblocking/AbstractHandler.java

```diff
@@ -98,11 +102,33 @@ final void unregisterOp(int operation) throws IOException {
         selectionKey.interestOps(selectionKey.interestOps() & ~operation);
     }
 
+    @SuppressWarnings({"checkstyle:npathcomplexity"})
     @Override
     public void onFailure(Throwable e) {
         if (e instanceof OutOfMemoryError) {
             ioService.onOutOfMemory((OutOfMemoryError) e);
         }
+
+        // if our selection key was cancelled due to replacement of the selector
+        // (see NonBlockingIOThread#selectWorkaround), then reschedule this handler for execution
+        if (e instanceof CancelledKeyException && selectionKey != null
+                && selectionKey.selector() != ioThread.getSelector()) {
+            // the selector was replaced, so obtain a new key for the new selector
```

Comment:
If a migration is in progress, you can't get any events since your selection key is closed.